### PR TITLE
#35 global and project role

### PR DESCRIPTION
**Pull request by Evgentik**

### CODE REVIEW

File: Recrutify/Recrutify.DataAccess/Repositories/Abstract/IUserRepository.cs [hunk 1]
```diff
@@ -10,5 +11,7 @@ public interface IUserRepository : IBaseRepository<User>
         Task<User> GetByEmailAsync(string email);
 
         Task<Dictionary<Guid, string>> GetNamesByIdsAsync(IEnumerable<Guid> ids);
+
+        Task<IEnumerable<User>> GetByRoles(List<Role> roles);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/Abstract/IUserRepository.cs:15]: pass IEnumerable<Role> to be in one style

File: Recrutify/Recrutify.Services/Services/UserService.cs [hunk 2]
```diff
@@ -29,12 +29,13 @@ public UserService(IUserRepository userRepository, IMapper mapper)
...
         public async Task<StaffGroupDTO> GetStaffByRolesAsync(List<Role> roles)
         {
-            var users = await _userRepository.GetAllAsync();
+            var users = await _userRepository.GetByRoles(roles);
 
             var staff = _mapper.Map<Dictionary<Role, List<StaffDTO>>>(
-                        users.SelectMany(p => p.ProjectRoles[Constants.GlobalProject.GlobalProjectId], (user, role) => new { user, role })
-                              .GroupBy(x => x.role)
-                              .ToDictionary(k => k.Key, i => i.Select(b => b.user).ToList()));
+                        users.AsEnumerable()
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/UserService.cs:35]: can delete AsEnumerable

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs [hunk 3]
```diff
@@ -35,5 +37,18 @@ public Task<User> GetByEmailAsync(string email)
...
+        {
+            var filter = Builders<BsonDocument>
+                            .Filter
+                            .ElemMatch(
+                                "ProjectRoles",
+                                Builders<BsonDocument>.Filter
+                                    .And(Builders<BsonDocument>.Filter.Eq("k", Constants.GlobalProject.GlobalProjectId)) &
+                                         Builders<BsonDocument>.Filter.AnyIn("v", roles));
+            var users = await GetBsonDocumentCollection().Find(filter).ToListAsync();
+            return BsonSerializer.Deserialize<List<User>>($"[{string.Join(",", users)}]");
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:51]: IEnumerable<User>

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs [hunk 4]
```diff
@@ -35,5 +37,18 @@ public Task<User> GetByEmailAsync(string email)
             var filter = _filterBuilder.Eq(u => u.Email, email);
             return GetCollection().Find(filter).FirstOrDefaultAsync();
         }
+
+        public async Task<IEnumerable<User>> GetByRoles(List<Role> roles)
+        {
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:42]: can write var filterBuilder = Builders<BsonDocument>
                            .Filter;
and use in other places

File: Recrutify/Recrutify.Services/Services/UserService.cs (hunk 2 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/UserService.cs]: +

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs (hunk 3 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:51]: +

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs (hunk 4 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:42]: +

File: Recrutify/Recrutify.DataAccess/GlobalSuppressions.cs [hunk 5]
```diff
@@ -6,3 +6,4 @@
 using System.Diagnostics.CodeAnalysis;
 
 [assembly: SuppressMessage("StyleCop.CSharp.MaintainabilityRules", "SA1401:Fields should be private", Justification = "<Ожидание>", Scope = "member", Target = "~F:Recrutify.DataAccess.Repositories.Abstract.BaseRepository`1._filterBuilder")]
+[assembly: SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1118:Parameter should not span multiple lines", Justification = "<Ожидание>", Scope = "member", Target = "~M:Recrutify.DataAccess.Repositories.UserRepository.GetByRoles(System.Collections.Generic.List{Recrutify.DataAccess.Models.Role})~System.Threading.Tasks.Task{System.Collections.Generic.IEnumerable{Recrutify.DataAccess.Models.User}}")]
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/GlobalSuppressions.cs:9]: remove it

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs (hunk 3 above)
Comment (elizavetapavliv) [reply Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:51]: try using ToJson

File: Recrutify/Recrutify.DataAccess/Repositories/Abstract/IUserRepository.cs (hunk 1 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/Abstract/IUserRepository.cs]: +

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs [hunk 6]
```diff
@@ -35,5 +38,18 @@ public Task<User> GetByEmailAsync(string email)
...
             return GetCollection().Find(filter).FirstOrDefaultAsync();
         }
+
+        public async Task<IEnumerable<User>> GetByRoles(IEnumerable<Role> roles)
+        {
+            var filterBuilder = Builders<BsonDocument>.Filter;
+            var filter = filterBuilder
+                            .ElemMatch(
+                                nameof(User.ProjectRoles),
+                                Builders<BsonDocument>.Filter
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:48]: filterBuilder

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs [hunk 7]
```diff
@@ -35,5 +38,18 @@ public Task<User> GetByEmailAsync(string email)
...
+                                    .And(Builders<BsonDocument>.Filter.Eq("k", Constants.GlobalProject.GlobalProjectId)) &
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:49]: filterBuilder

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs [hunk 8]
```diff
@@ -35,5 +38,18 @@ public Task<User> GetByEmailAsync(string email)
...
+                                         Builders<BsonDocument>.Filter.AnyIn("v", roles));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:50]: filterBuilder

File: Recrutify/Recrutify.DataAccess/GlobalSuppressions.cs (hunk 5 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/GlobalSuppressions.cs]: +

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs (hunk 6 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:48]: ?

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs (hunk 8 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:50]: +

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs (hunk 7 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:49]: +

File: Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs (hunk 6 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/UserRepository.cs:48]: +

**Review by elizavetapavliv — APPROVED** [review]

**Pull request merged**